src/Microsoft.ApplicationInsights.AspNetCore/TelemetryInitializers/TelemetryInitializerBase.cs

```diff
@@ -33,15 +33,18 @@ public void Initialize(ITelemetry telemetry)
                 return;
             }
 
-            var request = context.Features.Get<RequestTelemetry>();
-
-            if (request == null)
+            lock (context)
             {
-                AspNetCoreEventSource.Instance.LogTelemetryInitializerBaseInitializeRequestNull();
-                return;
-            }
+                var request = context.Features.Get<RequestTelemetry>();
 
-            this.OnInitializeTelemetry(context, request, telemetry);
+                if (request == null)
+                {
+                    AspNetCoreEventSource.Instance.LogTelemetryInitializerBaseInitializeRequestNull();
+                    return;
+                }
+
+                this.OnInitializeTelemetry(context, request, telemetry);
```

Comment:
potentially we can reduce the lock down to the immediate call to `HttpContext`. Most of telemetry initializers are not doing anything when the property is already set. So we wasting the `lock` on every cached item